### DISCUSSION TITLE
issue: isDirty reporting as false when using debounced async validation

### DISCUSSION THREAD

**Discussion by parkm** [react-hook-form/react-hook-form#12184] [Q&A]
### Version Number

7.53.0

### Codesandbox/Expo snack

https://codesandbox.io/p/sandbox/react-hook-form-isdirty-issue-f9myrm

### Steps to reproduce

In the codesandbox:
1. Delete the "test" (defaultValue) text on the input and replace it with "existing" (an error should popup indicating that validation occurred), notice that `isDirty` is `false`
2. Replace the text with "testa" and notice that `isDirty` is now `true`
3. At this point any value other than "test" will report `isDirty` as `true` which I believe is what is expected
4. If you then replace the text with the defaultValue "test" the bug will then reappear and will only be resolved when `isDirty` flips back to `true` at least once. 

So far I've only been able to get `isDirty` to be `true` when inputting some variation of "test" like "testa", "tes". However I'm seeing some inconsistencies with that where sometimes "testa" still shows `isDirty` as `false`

### Expected behaviour

Any value other than the defaultValue should mark `isDirty` as `true`.

For more context I'm trying to create a hook to be used for async api validations. Since right now anytime `isValid` is being used any changes to any other form inputs will result in the validation running again, even when the value it's validating has remained unchanged. So the hook will just return the previous validation result if the value was the same as before since the last validation. 

It looks like the validation is working fine, but something about it seems to be affecting how `isDirty` is being set. I'm not sure if there's a bug in this custom hook that would be causing this behavior, since the validation part is working and the form value is updating correctly. 

### What browsers are you seeing the problem on?

Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
I think it's related to your implementation on your custom input.

**Reply by parkm** [reply]
@bluebill1049 Here is a simplified version that doesn't use debouncing but still uses async validation
https://codesandbox.io/p/sandbox/react-hook-form-isdirty-issue-forked-j729lj

If I type `testasdf` `isDirty` is `false`
If I type `existing` `isDirty` is `false` and the "Item already exists" validation error pops up as expected
If I type `tes` `isDirty` is `true`

Do you have any insight on how this could be happening? Or do you have a recommendation on how to handle async validations?

**Reply by parkm** [reply]
@bluebill1049 Have you been able to take a look at the simplified example? 


```
    <Controller
      name={fieldName}
      control={control}
      defaultValue=""
      render={({ field, fieldState }) => (
        <>
          <input {...field} placeholder={label} />
          {fieldState.error && <p>{fieldState.error.message}</p>}
        </>
      )}
      rules={{
        validate: (value) => {
          return new Promise((resolve) => {
            setTimeout(() => {
              if (value === "existing") {
                resolve("Item already exists");
              } else {
                resolve(true);
              }
            }, 1000);
          });
        },
      }}
    />
    
```
    



That's essentially all there is yet I'm running into that issue with isDirty not updating as you'd expect. Do you have any suggestions on fixes or recommendations on implementing async validation?
